market/ipni/chunker/initial-chunker.go

```diff
@@ -0,0 +1,267 @@
+package chunker
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/multiformats/go-multihash"
+	"github.com/yugabyte/pgx/v5"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/market/ipni/ipniculib"
+)
+
+const longChainThreshold = 500_000
+
+// InitialChunker is used for initial entry chain creation.
+// It employs a dual strategy, where it tracks the number of entries, and:
+// For chains with less than longChainThreshold entries, it accumulates entries in memory.
+// which will then form a sorted chain of schema.EntryChunk nodes. This allows creation of
+// ad chains which are read from the index database.
+// For chains with more than longChainThreshold entries, it creates a chain of schema.EntryChunk nodes
+// in car order, noting where each chunk starts and ends in the car file.
+type InitialChunker struct {
+	chunkSize int
+
+	ingestedSoFar int64
+
+	// db-order ingest, up to longChainThreshold
+	dbMultihashes []multihash.Multihash
+
+	// car-order ingest, after longChainThreshold
+	carPending    []multihash.Multihash
+	carChunkStart *int64
+
+	prevChunks []carChunkMeta
+}
+
+type carChunkMeta struct {
+	link  ipld.Link
+	start int64
+	nodes int64
+}
+
+func NewInitialChunker() *InitialChunker {
```

Comment:
Can we init this with preallocated size/memory for carPending? it might be more efficient.

Reply:
Won't change that much, plus I intend to have this eventually rewritten to use cbor-gen everywhere anyways. Right now I'd guess that 99% of overhead will be in ipld-prime serialization/deserialization code.